File: sdks/ruby/calculator_sdk.rb
# Calculator API SDK for Ruby
# Full-featured Ruby SDK with error handling and caching

require 'net/http'
require 'json'
require 'uri'

class CalculatorSDK
  class Error < StandardError; end
  class APIError < Error
    attr_reader :status_code, :response_data
    
    def initialize(message, status_code = nil, response_data = nil)
      super(message)
      @status_code = status_code
      @response_data = response_data
    end
  end

  def initialize(api_key, base_url: 'https://api.yourcalculatorsite.com/v2', timeout: 30, retries: 3, debug: false)
    @api_key = api_key
    @base_url = base_url
    @timeout = timeout
    @retries = retries
    @debug = debug
    @cache = {}
  end

  def get_calculators(options = {})
    query_string = options.empty? ? '' : "?#{URI.encode_www_form(options)}"
    request('GET', "/calculators#{query_string}")
  end

  def calculate(input)
    data = input.is_a?(Hash) ? input : input.to_h
    request('POST', '/calculate', data)
  end

  private

  def request(method, endpoint, data = nil)
    uri = URI("#{@base_url}#{endpoint}")
    
    @retries.times do |attempt|
      begin
        http = Net::HTTP.new(uri.host, uri.port)
        http.use_ssl = uri.scheme == 'https'
        http.read_timeout = @timeout

        case method
        when 'GET'
          request = Net::HTTP::Get.new(uri)
        when 'POST'
          request = Net::HTTP::Post.new(uri)
          request.body = data.to_json if data
        end

        request['Authorization'] = "Bearer #{@api_key}"
        request['Content-Type'] = 'application/json'
        request['User-Agent'] = 'CalculatorSDK-Ruby/2.0.0'

        puts "[CalculatorSDK] #{method} #{uri}" if @debug

        response = http.request(request)

        unless response.is_a?(Net::HTTPSuccess)
          error_data = JSON.parse(response.body) rescue {}
          error_message = error_data['error'] || response.message
          raise APIError.new("API Error #{response.code}: #{error_message}", response.code.to_i, error_data)
        end

        result = JSON.parse(response.body)
        puts "[CalculatorSDK] Response: #{result}" if @debug
        
        return result

      rescue => error
        if attempt < @retries - 1
          delay = 2 ** attempt
          puts "[CalculatorSDK] Attempt #{attempt + 1} failed, retrying in #{delay}s: #{error.message}" if @debug
          sleep(delay)
        else
          raise error
        end
      end
    end
  end
end

# Example usage:
# sdk = CalculatorSDK.new('your_api_key', debug: true)
# calculators = sdk.get_calculators(category: 'finance')
# result = sdk.calculate(calculatorId: 'mortgage', inputs: { principal: 300000, rate: 3.5, term: 30 })

File: sdks/javascript/calculator-sdk.ts
/**
 * Calculator API SDK for JavaScript/TypeScript
 * Full-featured SDK with TypeScript support, caching, and error handling
 */

export interface CalculatorSDKConfig {
 apiKey: string;
 baseUrl?: string;
 timeout?: number;
 retries?: number;
 cacheEnabled?: boolean;
 debug?: boolean;
}

export interface CalculationInput {
 calculatorId: string;
 inputs: Record<string, any>;
 format?: 'simple' | 'detailed' | 'breakdown';
 precision?: number;
 currency?: string;
 locale?: string;
}

export interface CalculationResult {
 result: any;
 formatted?: any;
 breakdown?: any[];
 metadata: {
 duration: number;
 cached: boolean;
 precision: number;
 currency: string;
 locale: string;
 };
}

export class CalculatorSDK {
 private config: Required<CalculatorSDKConfig>;
 private cache = new Map<string, { data: any; expires: number }>();

 constructor(config: CalculatorSDKConfig) {
 this.config = {
 baseUrl: 'https://api.yourcalculatorsite.com/v2',
 timeout: 30000,
 retries: 3,
 cacheEnabled: true,
 debug: false,
 ...config
 };
 }

 /**
 * Get list of available calculators
 */
 async getCalculators(options: {
 category?: string;
 featured?: boolean;
 search?: string;
 page?: number;
 limit?: number;
 } = {}): Promise<any> {
 const params = new URLSearchParams();
 Object.entries(options).forEach(([key, value]) => {
 if (value !== undefined) {
 params.append(key, value.toString());
 }
 });

 return this.request('GET', `/calculators?${params}`);
 }

 /**
 * Perform a calculation
 */
 async calculate(input: CalculationInput): Promise<CalculationResult> {
 // Generate cache key for calculation
 const cacheKey = this.generateCalculationCacheKey(input);
 
 if (this.config.cacheEnabled) {
 const cached = this.getFromCache(cacheKey);
 if (cached) {
 cached.metadata.cached = true;
 return cached;
 }
 }

 const result = await this.request('POST', '/calculate', input);
 
 if (this.config.cacheEnabled) {
 this.setCache(cacheKey, result, 3600000); // Cache for 1 hour
 }

 return result;
 }

 /**
 * Generic request method with retry logic
 */
 private async request(
 method: string, 
 endpoint: string, 
 data?: any
 ): Promise<any> {
 const url = `${this.config.baseUrl}${endpoint}`;
 let lastError: Error;

 for (let attempt = 0; attempt <= this.config.retries; attempt++) {
 try {
 const controller = new AbortController();
 const timeoutId = setTimeout(() => controller.abort(), this.config.timeout);

 const requestOptions: RequestInit = {
 method,
 headers: {
 'Content-Type': 'application/json',
 'Authorization': `Bearer ${this.config.apiKey}`,
 'User-Agent': 'CalculatorSDK/2.0.0'
 },
 signal: controller.signal
 };

 if (data && method !== 'GET') {
 requestOptions.body = JSON.stringify(data);
 }

 if (this.config.debug) {
 console.log(`${method} ${url}`, data);
 }

 const response = await fetch(url, requestOptions);
 clearTimeout(timeoutId);

 if (!response.ok) {
 const error = await response.json().catch(() => ({}));
 throw new Error(`API Error ${response.status}: ${error.error || response.statusText}`);
 }

 const result = await response.json();

 if (this.config.debug) {
 console.log('Response:', result);
 }
 
 return result;

 } catch (error) {
 lastError = error as Error;
 
 if (attempt < this.config.retries) {
 const delay = Math.pow(2, attempt) * 1000; // Exponential backoff
 if (this.config.debug) {
 console.log(`Attempt ${attempt + 1} failed, retrying in ${delay}ms:`, error.message);
 }
 await this.sleep(delay);
 }
 }
 }

 throw lastError!;
 }

 private getFromCache(key: string): any | null {
 const cached = this.cache.get(key);
 if (cached && cached.expires > Date.now()) {
 return cached.data;
 }
 if (cached) {
 this.cache.delete(key);
 }
 return null;
 }

 private setCache(key: string, data: any, ttl: number): void {
 this.cache.set(key, {
 data,
 expires: Date.now() + ttl
 });
 }

 private generateCalculationCacheKey(input: CalculationInput): string {
 const key = {
 calculatorId: input.calculatorId,
 inputs: input.inputs,
 format: input.format || 'simple',
 precision: input.precision || 2,
 currency: input.currency || 'USD',
 locale: input.locale || 'en-US'
 };
 return `calc:${btoa(JSON.stringify(key))}`;
 }

 private sleep(ms: number): Promise<void> {
 return new Promise(resolve => setTimeout(resolve, ms));
 }
}

// Default export
export default CalculatorSDK;
